### PR TITLE
scripts and compose files for integration testing

### PR DESCRIPTION
**Pull request by tiainen**
## Description

Fixes pyrsia/pyrsia#234

This PR adds scripts and the necessary docker files to start an integration test of the pyrsia node. To start the integration test, navigate to the `tests` directory and run: `./pyrsia-it-test.sh`.

It mainly works by leveraging docker compose and dind (docker-in-docker) containers. The docker compose consists of the following services:

 * two dind containers, where each registry mirror is configured to a pyrsia node
 * two pyrsia nodes
 * a bare debian container that runs the integration test

The dind containers are started up first. Afterwards we start up pyrsia node 1 and then pyrsia node 2. Finally, the integration test container is started which runs the script located at `tests/pyrsia-it.sh`. The script itself pulls the `alpine:3.15.4` image from the first docker container. Because of the registry mirror configuration, it downloads the image from pyrsia node 1. Since it is the first time that the image is requested, it is not yet available locally and therefore will be downloaded from docker.io. Then the script will pull the alpine image from the second docker container. Because of the registry mirror configuration, the download will happen from pyrsia node 2. This time the image is available in the pyrsia network (where pyrsia node 1 is a provider of the image) and hence the image is downloaded over the libp2p network from pyrsia node 1 instead of from docker.io.

Finally, the integration test script will verify the logs of the second pyrsia node to verify that the images was indeed succesfully downloaded over the libp2p network.

## PR Checklist

- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/good_pr.md)
- [x] I've included a good title and brief description along with how to review them.
- [x] I've linked any associated an [issue](https://github.com/pyrsia/pyrsia/issues).
- [ ] I've requested a review  from `pyrsia/collaborators`.

### Code Contributions

- [x] I've built the code `cargo build --all-targets` successfully.
- [x] I've run the unit tests `cargo test --workspace` and everything passes.
- [x] I've made sure my rust toolchain is current `rustup update`.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -43,3 +43,26 @@ RUN <<EOT bash
     rm -rf /var/lib/apt/lists/*
 EOT
 COPY --from=dbuild /out/pyrsia_node /usr/local/bin/
+
+FROM debian:buster-slim AS node-it
+ARG P2P_KEYPAIR
+ENTRYPOINT ["pyrsia_node"]
+ENV RUST_LOG=info
+RUN <<EOT bash
```
Comment (chb0github) [inline Dockerfile:51]: mmm, normaly this is done

```
RUN apt-get update && apt-get install -y ca-certificates &&...
```

A heredoc might make sense if you're trying to do variable substitution, which you're not

File: Dockerfile [hunk 2]
```diff
@@ -43,3 +43,26 @@ RUN <<EOT bash
...
+    rm -rf /var/lib/apt/lists/*
+    mkdir /pyrsia
+EOT
+COPY tests/${P2P_KEYPAIR} /pyrsia/p2p_keypair.ser
+COPY --from=dbuild /out/pyrsia_node /usr/local/bin/
+
+FROM debian:buster-slim AS it-test
+RUN apt update && apt install -y apt-transport-https ca-certificates curl gnupg2 software-properties-common
+RUN curl -fsSL https://download.docker.com/linux/debian/gpg | gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg
+RUN echo "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/debian $(lsb_release -cs) stable" | tee /etc/apt/sources.list.d/docker.list
```
Comment (chb0github) [inline Dockerfile:65]: multi-line this please

File: tests/docker-compose-it.yml [hunk 3]
```diff
@@ -0,0 +1,82 @@
+services:
+  docker1:
+    image: docker:dind
```
Comment (chb0github) [inline tests/docker-compose-it.yml:3]: Help me understand why this is needed. I mean, even docker recommends against this

File: tests/docker-compose-it.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    environment:
+      - RUST_LOG=pyrsia=debug
+      - DEV_MODE=on
+    command: ["--host" , "0.0.0.0", "--listen", "/ip4/0.0.0.0/tcp/44002", "--peer", "/dns4/pyrsia1/tcp/44001/p2p/12D3KooWS2qJtetLCeerZg5hNZRLpaCcP7bF4ArpT2FaLXAEL1dL"]
+    networks:
+      - docker2
+      - pyrsia
+
+  pyrsia-it:
+    image: pyrsia/node-it
```
Comment (chb0github) [inline tests/docker-compose-it.yml:63]: what is "pyrsia/node-it`

File: tests/docker-compose-it.yml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      target: it-test
+    volumes:
+      - certs-client1:/certs/client1:ro
+      - certs-client2:/certs/client2:ro
+    networks:
+      - docker1
+      - docker2
+      - pyrsia
+
+networks:
```
Comment (chb0github) [inline tests/docker-compose-it.yml:75]: Why are they all on different networks?

File: tests/pyrsia-it-test.sh [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+docker-compose -f docker-compose-it.yml -p pyrsia-it down
+docker volume rm pyrsia-it_certs-client1 pyrsia-it_certs-client2
+
+echo "Starting two docker daemons"
+docker-compose -f docker-compose-it.yml -p pyrsia-it up -d --build docker1 docker2
+echo "Waiting for docker daemons"
+w1=0
+while [ $w1 -ne 2 ]; do
+    sleep 5
+    w1=$(docker-compose -f docker-compose-it.yml -p pyrsia-it logs docker1 docker2 | grep -F "API listen on [::]:2376" | wc -l)
```
Comment (chb0github) [inline tests/pyrsia-it-test.sh:13]: Hmm, you know that if there is an error, this will never actually end, right?

File: tests/pyrsia-it-test.sh [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+done
+
+echo "Starting pyrsia node 1"
+docker-compose -f docker-compose-it.yml -p pyrsia-it up -d --build pyrsia1
+echo "Waiting for pyrsia node 1"
+w2=0
+while [ $w2 -ne 1 ]; do
+    sleep 5
+    w2=$(docker-compose -f docker-compose-it.yml -p pyrsia-it logs pyrsia1 | grep -F "> Listen for p2p events" | wc -l)
```
Comment (chb0github) [inline tests/pyrsia-it-test.sh:22]: I am not sure where this text comes from, but it's arbitrary log output from code which could change and you'd not know - and then this loop would go forever

File: tests/pyrsia-it-test.sh [hunk 8]
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+
```
Comment (chb0github) [inline tests/pyrsia-it-test.sh:2]: `set -e` always

File: tests/pyrsia-it-test.sh [hunk 9]
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
```
Comment (chb0github) [inline tests/pyrsia-it-test.sh:1]: you should be using `#!/usr/bin/env bash` so as to pickup bash from wherever it's installed - make it more portable

File: tests/pyrsia-it-test.sh [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+while [ $w3 -ne 1 ]; do
+    sleep 5
+    w3=$(docker-compose -f docker-compose-it.yml -p pyrsia-it logs pyrsia2 | grep -F "> Listen for p2p events" | wc -l)
+done
+
+echo "Running integration test"
+docker-compose -f docker-compose-it.yml -p pyrsia-it up --build pyrsia-it
+
+echo "Validating integration test"
+v1=$(docker-compose -f docker-compose-it.yml -p pyrsia-it logs pyrsia2 | grep "Step 1: NO, alpine/3.15.4 does not exist in the metadata manager\\." | wc -l)
```
Comment (chb0github) [inline tests/pyrsia-it-test.sh:38]: This is very opaque. If you're going to do this, then it should be:
1. run code
2. assert result

Not, run 10 tests and assert all 10. You want to know which failed and when. I have no idea what the different is between `v6` and `v9`

**Review by chb0github — CHANGES_REQUESTED** [review]
There are definitely ways to improve

**Review by fishseabowl — APPROVED** [review]
LGTM

**Comment by tiainen** [comment]
Closing in favor of the work done here: https://github.com/pyrsia/pyrsia-integration-tests

**Pull request closed without merging**